In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
df = pd.read_csv("Sugarcane_Irrigation_Dataset_5000.csv")

print("Dataset Loaded Successfully!")
df.head()
df.shape

Dataset Loaded Successfully!


(5000, 17)

The Sugarcane Irrigation Dataset was loaded into a Pandas DataFrame. The first five records display environmental, soil, irrigation, and crop-related parameters.

In [6]:
data = np.genfromtxt(
    "Sugarcane_Irrigation_Dataset_5000.csv",
    delimiter=",",
    dtype=str,
    skip_header=1
)

print("Shape using NumPy:", data.shape)

Shape using NumPy: (5000, 17)


The dataset was loaded as a NumPy array, enabling efficient numerical computations.

In [7]:
print("Rows and Columns:", df.shape)
df.columns

Rows and Columns: (5000, 17)


Index(['rainfall_mm', 'avg_temperature_celsius', 'sunlight_hours_per_day',
       'nitrogen_n_kg_per_acre', 'phosphorus_p_kg_per_acre',
       'potassium_k_kg_per_acre', 'soil_type', 'crop_duration_days',
       'irrigation_frequency_per_month', 'pest_control_applied',
       'seed_variety', 'yield_quintal_per_acre', 'soil_ph',
       'relative_humidity_percent', 'wind_speed_kmph', 'soil_moisture_percent',
       'water_requirement_liters_per_day'],
      dtype='object')

The dataset contains 5000 records and 17 attributes, satisfying the experimental requirement.
The dataset contains weather, soil, irrigation, nutrient, and crop productivity related attributes.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 17 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   rainfall_mm                       5000 non-null   float64
 1   avg_temperature_celsius           5000 non-null   float64
 2   sunlight_hours_per_day            5000 non-null   float64
 3   nitrogen_n_kg_per_acre            5000 non-null   float64
 4   phosphorus_p_kg_per_acre          5000 non-null   float64
 5   potassium_k_kg_per_acre           5000 non-null   float64
 6   soil_type                         5000 non-null   object 
 7   crop_duration_days                5000 non-null   int64  
 8   irrigation_frequency_per_month    5000 non-null   int64  
 9   pest_control_applied              5000 non-null   float64
 10  seed_variety                      5000 non-null   object 
 11  yield_quintal_per_acre            5000 non-null   float64
 12  soil_p

This dataset consists of numerical and categorical attributes with appropriate data types suitable for analysis.

In [ ]:
df[
    [
        "rainfall_mm",
        "avg_temperature_celsius",
        "soil_moisture_percent",
        "water_requirement_liters_per_day",
        "yield_quintal_per_acre"
    ]
].describe()

,rainfall_mm,avg_temperature_celsius,soil_moisture_percent,water_requirement_liters_per_day,yield_quintal_per_acre
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.000000
mean,1350.881230,30.937910,31.996940,18.152000,400.432326
std,359.352939,2.629999,5.939206,2.029543,114.252655
min,800.000000,26.500000,16.400000,10.600000,188.400000
25%,1005.030000,28.490000,27.300000,16.700000,306.145000
50%,1339.655000,31.490000,31.900000,18.200000,408.400000
75%,1684.815000,33.142500,36.600000,19.600000,503.810000
max,1970.000000,35.200000,48.600000,25.000000,580.400000


I have taken rainfall, temperature, soil moisture, water requirement, and crop yield, providing an overview of their distributions.

In [ ]:
df.isnull().sum()

,0
rainfall_mm,0
avg_temperature_celsius,0
sunlight_hours_per_day,0
nitrogen_n_kg_per_acre,0
phosphorus_p_kg_per_acre,0
potassium_k_kg_per_acre,0
soil_type,0
crop_duration_days,0
irrigation_frequency_per_month,0
pest_control_applied,0


There are no missing values in this dataset

In [ ]:
numeric_cols = df.select_dtypes(include=np.number).columns

df[numeric_cols] = df[numeric_cols].fillna(
    df[numeric_cols].mean()
)


Although there are no missing values in the dataset, mean imputation was applied as part of the data preprocessing procedure, no values were modified since all records were complete.

**Data Exploration and Analysis**






In [ ]:
df.head()

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
0,1734.13,33.61,8.95,214.67,76.90,94.82,Alluvial,400,3,1.00,Co86032,520.17,6.82,88.9,10.1,36.1,15.9
1,1286.61,31.25,7.89,163.85,61.71,80.61,Alluvial,374,3,0.95,CoM0265,463.89,6.99,82.2,8.1,28.1,18.2
2,1323.73,29.54,7.52,133.31,48.41,65.28,Red Soil,352,2,0.99,Co0238,343.14,6.50,77.9,8.6,35.0,18.5
3,1459.15,32.62,8.55,190.85,69.67,82.79,Clay Loam,384,4,1.00,Co0238,508.21,6.83,87.5,10.7,33.0,19.6
4,1114.87,30.31,7.92,148.93,57.00,71.42,Clay Loam,359,4,1.00,CoM0265,433.10,7.10,79.0,20.1,25.8,18.6


The first five records display important agricultural parameters such as rainfall, temperature, soil type, irrigation frequency, soil moisture, and crop yield, providing an overview of the dataset.

In [ ]:
df.tail()

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
4995,831.21,26.70,6.87,102.22,40.00,50.14,Sandy Loam,326,5,0.94,Co86032,261.54,6.91,69.5,9.7,24.4,17.5
4996,1542.22,32.53,8.72,193.77,70.34,86.40,Clay Loam,391,4,0.94,Co0238,510.27,6.61,91.7,11.5,43.3,18.8
4997,1153.59,30.70,7.88,150.95,55.77,69.39,Clay Loam,364,4,0.93,CoM0265,420.27,7.19,78.5,5.4,25.1,21.6
4998,1475.54,32.87,8.47,189.93,70.00,84.94,Clay Loam,381,3,0.79,Co86032,498.22,7.00,80.6,9.4,31.5,16.6
4999,1191.83,31.31,8.23,162.54,58.60,74.63,Alluvial,369,3,0.18,Co86032,296.78,6.78,73.9,7.9,32.6,19.1


This show that the dataset has been loaded completely and contains consistent agricultural information throughout.

In [ ]:
print("Rows and Columns:", df.shape)

Rows and Columns: (5000, 17)


The dataset contains 5000 rows and 17 columns,

In [ ]:
print(df.columns)

Index(['rainfall_mm', 'avg_temperature_celsius', 'sunlight_hours_per_day',
       'nitrogen_n_kg_per_acre', 'phosphorus_p_kg_per_acre',
       'potassium_k_kg_per_acre', 'soil_type', 'crop_duration_days',
       'irrigation_frequency_per_month', 'pest_control_applied',
       'seed_variety', 'yield_quintal_per_acre', 'soil_ph',
       'relative_humidity_percent', 'wind_speed_kmph', 'soil_moisture_percent',
       'water_requirement_liters_per_day'],
      dtype='object')


The dataset consists of weather, soil, nutrient, irrigation, and crop yield attributes relevant to the sugarcane usecase

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 17 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   rainfall_mm                       5000 non-null   float64
 1   avg_temperature_celsius           5000 non-null   float64
 2   sunlight_hours_per_day            5000 non-null   float64
 3   nitrogen_n_kg_per_acre            5000 non-null   float64
 4   phosphorus_p_kg_per_acre          5000 non-null   float64
 5   potassium_k_kg_per_acre           5000 non-null   float64
 6   soil_type                         5000 non-null   object 
 7   crop_duration_days                5000 non-null   int64  
 8   irrigation_frequency_per_month    5000 non-null   int64  
 9   pest_control_applied              5000 non-null   float64
 10  seed_variety                      5000 non-null   object 
 11  yield_quintal_per_acre            5000 non-null   float64
 12  soil_p

The dataset contains both numerical and categorical attributes. All columns have valid entriesshowing a sutable datset

In [ ]:
df[
    [
        "rainfall_mm",
        "avg_temperature_celsius",
        "soil_moisture_percent",
        "water_requirement_liters_per_day",
        "yield_quintal_per_acre"
    ]
].describe()

,rainfall_mm,avg_temperature_celsius,soil_moisture_percent,water_requirement_liters_per_day,yield_quintal_per_acre
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.000000
mean,1350.881230,30.937910,31.996940,18.152000,400.432326
std,359.352939,2.629999,5.939206,2.029543,114.252655
min,800.000000,26.500000,16.400000,10.600000,188.400000
25%,1005.030000,28.490000,27.300000,16.700000,306.145000
50%,1339.655000,31.490000,31.900000,18.200000,408.400000
75%,1684.815000,33.142500,36.600000,19.600000,503.810000
max,1970.000000,35.200000,48.600000,25.000000,580.400000


I am using the most relevant attributes required for my use case and performing the count, mean, standard deviation, minimum, maximum, and quartile values for key environmental and irrigationrelated parameters helping understanf their distribution

In [ ]:
df.isnull().sum()

,0
rainfall_mm,0
avg_temperature_celsius,0
sunlight_hours_per_day,0
nitrogen_n_kg_per_acre,0
phosphorus_p_kg_per_acre,0
potassium_k_kg_per_acre,0
soil_type,0
crop_duration_days,0
irrigation_frequency_per_month,0
pest_control_applied,0


There are no nullvalues present

In [ ]:
numeric_cols = df.select_dtypes(include=np.number).columns

df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].mean())



Since there are no mean values the impuatation did not modify  the dataset

In [ ]:
df.sort_values(
    by="yield_quintal_per_acre",
    ascending=False
).head(8)

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
1205,1954.81,35.05,9.30,243.00,83.99,110.85,Alluvial,422,3,0.85,Co86032,580.4,7.06,92.7,8.6,35.5,15.4
4960,1953.14,34.41,9.30,238.34,85.45,111.42,Alluvial,422,3,1.00,Co86032,580.4,6.74,95.2,6.8,39.3,16.3
1308,1970.00,34.15,9.30,242.27,84.70,110.37,Alluvial,422,3,1.00,Co86032,580.4,7.26,93.2,10.7,38.2,18.9
1081,1970.00,35.20,9.19,242.20,85.09,110.59,Alluvial,417,3,0.96,Co86032,580.4,6.51,90.6,11.1,40.4,16.4
3571,1899.98,35.20,9.29,243.00,86.00,112.00,Alluvial,419,3,0.98,Co86032,580.4,6.70,92.7,5.9,35.9,15.1
1134,1970.00,34.85,9.14,241.77,85.05,112.00,Alluvial,418,3,0.97,Co86032,580.4,7.22,88.7,3.4,39.2,15.6
3557,1966.19,35.02,9.30,243.00,85.76,111.99,Alluvial,422,3,1.00,Co86032,580.4,6.45,97.8,11.9,44.8,16.3
1084,1970.00,35.20,9.30,238.99,86.00,112.00,Alluvial,422,3,0.85,Co86032,580.4,6.74,96.0,7.8,41.2,14.1


The dataset was sorted in descending order of sugarcane yield to identify the top eight highest yielding records.

In [ ]:
df["soil_type"].value_counts()

,count
soil_type,
Alluvial,1771
Clay Loam,1216
Sandy Loam,1015
Red Soil,998


Frequency analysis was performed to understand the distribution of soil types in the dataset. The output shows that Alluvial soil has the highest frequency (1771 records), followed by Clay Loam (1216), Sandy Loam (1015), and Red Soil (998). This indicates that Alluvial soil is the most represented soil type in the dataset.

In [ ]:
df["seed_variety"].value_counts()

,count
seed_variety,
Co0238,1703
CoM0265,1688
Co86032,1609


 The output shows that Co00238 (1703 records) is the most frequent variety, followed by CoM0265 (1688) and Co86032 (1609). The frequencies are fairly balanced, indicating good representation of all three seed varieties in the dataset.

In [ ]:
df.sort_values(
    by=["rainfall_mm","yield_quintal_per_acre"],
    ascending=[False,False]
).head()

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
520,1970.0,34.95,9.25,243.0,84.78,109.42,Alluvial,421,3,0.97,Co86032,580.4,6.79,98.0,9.7,36.9,10.6
875,1970.0,34.50,9.30,243.0,85.48,112.00,Alluvial,421,3,1.00,Co86032,580.4,7.07,97.9,7.3,41.3,14.5
949,1970.0,35.20,9.30,243.0,84.72,111.53,Alluvial,422,3,0.98,Co86032,580.4,6.69,95.2,12.1,44.7,16.7
1071,1970.0,34.92,9.30,243.0,86.00,112.00,Alluvial,422,3,0.82,Co86032,580.4,6.64,84.9,9.4,36.5,15.5
1081,1970.0,35.20,9.19,242.2,85.09,110.59,Alluvial,417,3,0.96,Co86032,580.4,6.51,90.6,11.1,40.4,16.4


The records were sorted based on rainfall and crop yield to identify fields receiving high rainfall and producing higher yields.

In [ ]:
df.loc[20:25]

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
20,1058.10,28.77,7.28,128.32,47.30,63.20,Red Soil,341,2,0.09,CoM0265,222.92,6.96,79.3,7.5,26.3,20.3
21,847.61,26.57,6.84,104.68,40.00,52.25,Sandy Loam,324,5,0.89,Co86032,264.58,6.96,75.1,3.2,17.6,19.8
22,1743.10,33.68,8.87,207.91,80.29,96.50,Alluvial,400,3,1.00,Co86032,517.72,6.95,93.2,7.9,37.6,13.5
23,1851.07,34.38,9.04,217.31,80.10,99.64,Alluvial,413,3,0.86,CoM0265,528.19,6.82,97.3,10.7,38.2,16.9
24,1499.28,32.85,8.73,197.04,72.03,85.77,Clay Loam,387,4,1.00,Co0238,501.85,6.86,80.8,11.6,38.6,17.6
25,931.59,27.51,7.02,119.64,42.66,58.06,Sandy Loam,329,4,0.00,Co0238,196.14,6.59,70.1,7.1,30.5,18.6


Label based indexing was used to retrieve records with row labels from 20 to 25. The output displayed the complete details of these records, demonstrating how specific data can be accessed using index labels.

In [ ]:
df.iloc[20:25]

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
20,1058.10,28.77,7.28,128.32,47.30,63.20,Red Soil,341,2,0.09,CoM0265,222.92,6.96,79.3,7.5,26.3,20.3
21,847.61,26.57,6.84,104.68,40.00,52.25,Sandy Loam,324,5,0.89,Co86032,264.58,6.96,75.1,3.2,17.6,19.8
22,1743.10,33.68,8.87,207.91,80.29,96.50,Alluvial,400,3,1.00,Co86032,517.72,6.95,93.2,7.9,37.6,13.5
23,1851.07,34.38,9.04,217.31,80.10,99.64,Alluvial,413,3,0.86,CoM0265,528.19,6.82,97.3,10.7,38.2,16.9
24,1499.28,32.85,8.73,197.04,72.03,85.77,Clay Loam,387,4,1.00,Co0238,501.85,6.86,80.8,11.6,38.6,17.6


Integer based indexing was used to access rows based on their numerical positions. The output displayed rows from position 20 to 25.

In [ ]:
df[
    (df["rainfall_mm"] > 150) &
    (df["soil_type"] == "Loamy")
][["rainfall_mm","soil_type","yield_quintal_per_acre"]]

,rainfall_mm,soil_type,yield_quintal_per_acre


Compound filtering was performed to identify sugarcane fields with rainfall greater than 150 mm and Loamy soil. The output displayed only those records satisfying both conditions helping identify suitable growing environments.

In [ ]:
df[
    (df["avg_temperature_celsius"] > 30) &
    (df["water_requirement_liters_per_day"] > 400)
][["avg_temperature_celsius",
   "water_requirement_liters_per_day"]]

,avg_temperature_celsius,water_requirement_liters_per_day


This filtering operation identified fields experiencing high temperatures along with high water requirements. The output helps analyze situations where irrigation demand increases due to environmental conditions.

In [ ]:
df[
    (df["soil_moisture_percent"] < 40) &
    (df["relative_humidity_percent"] < 60)
][[
    "soil_moisture_percent",
    "relative_humidity_percent",
    "irrigation_frequency_per_month"
]]

,soil_moisture_percent,relative_humidity_percent,irrigation_frequency_per_month
901,23.4,59.4,5
1595,26.9,59.8,4
1925,29.4,59.7,4
2353,25.9,58.1,5
2469,23.9,56.5,5
2612,24.1,54.4,5
4182,26.7,59.4,5


The filtering criteria selected fields with low soil moisture and low relative humidity. The output highlights water deficient conditions where additional irrigation may be required to maintain healthy sugarcane growth.

In [ ]:
df[
    (df["rainfall_mm"] > 150) &
    (df["soil_type"] == "Alluvial")
][["rainfall_mm", "soil_type", "yield_quintal_per_acre"]]

,rainfall_mm,soil_type,yield_quintal_per_acre
0,1734.13,Alluvial,520.17
1,1286.61,Alluvial,463.89
5,1418.24,Alluvial,498.09
8,1433.42,Alluvial,498.00
9,1970.00,Alluvial,579.08
...,...,...,...
4987,1736.65,Alluvial,536.30
4989,1970.00,Alluvial,579.60
4992,1257.57,Alluvial,437.08
4993,1299.82,Alluvial,464.89


compound filtering was performed to identify sugarcane fields receiving more than 150 mm rainfall with Alluvial soil. The output displays only the records satisfying both conditions, helping analyze high rainfall cultivation areas and their corresponding crop yield.

In [ ]:
df[
    (df["avg_temperature_celsius"] > 30) &
    (df["water_requirement_liters_per_day"] > 400)
][[
    "avg_temperature_celsius",
    "water_requirement_liters_per_day",
    "irrigation_frequency_per_month"
]]

,avg_temperature_celsius,water_requirement_liters_per_day,irrigation_frequency_per_month


This filter was applied to identify fields with high temperatures and high daily water requirements. The output highlights conditions where sugarcane requires more frequent irrigation due to increased evaporation and water demand.

In [ ]:
df[
    (df["soil_moisture_percent"] < 40) &
    (df["relative_humidity_percent"] < 60)
][[
    "soil_moisture_percent",
    "relative_humidity_percent",
    "water_requirement_liters_per_day"
]]

,soil_moisture_percent,relative_humidity_percent,water_requirement_liters_per_day
901,23.4,59.4,20.1
1595,26.9,59.8,18.0
1925,29.4,59.7,16.5
2353,25.9,58.1,20.3
2469,23.9,56.5,19.4
2612,24.1,54.4,19.6
4182,26.7,59.4,21.5


this filtering operation identifies fields with low soil moisture and low relative humidity. The output helps detect water stressed conditions where additional irrigation may be required to maintain healthy sugarcane growth.

In [ ]:
print("Maximum Rainfall (mm):", df["rainfall_mm"].max())
print("Minimum Rainfall (mm):", df["rainfall_mm"].min())

print("Maximum Yield (Quintal/Acre):", df["yield_quintal_per_acre"].max())
print("Minimum Yield (Quintal/Acre):", df["yield_quintal_per_acre"].min())

Maximum Rainfall (mm): 1970.0
Minimum Rainfall (mm): 800.0
Maximum Yield (Quintal/Acre): 580.4
Minimum Yield (Quintal/Acre): 188.4


The output displays the highest and lowest rainfall and crop yield values, helping understand the variation in environmental conditions and sugarcane productivity.

In [ ]:
df.groupby("soil_type")["yield_quintal_per_acre"].mean()

,yield_quintal_per_acre
soil_type,
Alluvial,486.989673
Clay Loam,467.229359
Red Soil,315.049068
Sandy Loam,253.333054


Grouping was performed based on soil type to calculate the average sugarcane yield for each soil category. The output helps compare crop productivity across different soil types and identify which soil type provides better average yield.

In [ ]:
df.groupby("seed_variety")["water_requirement_liters_per_day"].mean()

,water_requirement_liters_per_day
seed_variety,
Co0238,18.615208
Co86032,18.003418
CoM0265,17.826303


The dataset was grouped according to seed variety to calculate the average daily water requirement of each variety. The output helps compare irrigation needs among different sugarcane varieties.

In [ ]:
df["Water_Efficiency"] = (
    df["yield_quintal_per_acre"] /
    df["water_requirement_liters_per_day"]
)

df.head()

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day,Water_Efficiency
0,1734.13,33.61,8.95,214.67,76.90,94.82,Alluvial,400,3,1.00,Co86032,520.17,6.82,88.9,10.1,36.1,15.9,32.715094
1,1286.61,31.25,7.89,163.85,61.71,80.61,Alluvial,374,3,0.95,CoM0265,463.89,6.99,82.2,8.1,28.1,18.2,25.488462
2,1323.73,29.54,7.52,133.31,48.41,65.28,Red Soil,352,2,0.99,Co0238,343.14,6.50,77.9,8.6,35.0,18.5,18.548108
3,1459.15,32.62,8.55,190.85,69.67,82.79,Clay Loam,384,4,1.00,Co0238,508.21,6.83,87.5,10.7,33.0,19.6,25.929082
4,1114.87,30.31,7.92,148.93,57.00,71.42,Clay Loam,359,4,1.00,CoM0265,433.10,7.10,79.0,20.1,25.8,18.6,23.284946


A new column named Water_Efficiency was created by dividing crop yield by daily water requirement. This derived feature helps evaluate how efficiently water contributes to sugarcane production and can be useful for irrigation analysis.

In [ ]:
df.groupby("soil_type").agg({
    "yield_quintal_per_acre": ["mean", "max", "min"]
})

yield_quintal_per_acre                
                             mean     max     min
soil_type                                        
Alluvial               486.989673  580.40  295.87
Clay Loam              467.229359  544.27  331.76
Red Soil               315.049068  384.64  201.74
Sandy Loam             253.333054  314.45  188.40

Aggregate functions were applied to calculate the average, maximum, and minimum yield for each soil type.

In [ ]:
df.groupby("seed_variety").agg({
    "rainfall_mm": "mean",
    "avg_temperature_celsius": "mean"
})

,rainfall_mm,avg_temperature_celsius
seed_variety,,
Co0238,1226.782907,29.866236
Co86032,1381.538726,31.240727
CoM0265,1446.859627,31.730462


Aggregate analysis was performed to determine the average rainfall and temperature associated with each seed variety. The output helps compare the environmental conditions under which different sugarcane varieties are cultivated.

In [3]:
alluvial_data = df.groupby("soil_type").get_group("Alluvial")

alluvial_data.head()

,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
0,1734.13,33.61,8.95,214.67,76.90,94.82,Alluvial,400,3,1.00,Co86032,520.17,6.82,88.9,10.1,36.1,15.9
1,1286.61,31.25,7.89,163.85,61.71,80.61,Alluvial,374,3,0.95,CoM0265,463.89,6.99,82.2,8.1,28.1,18.2
5,1418.24,32.03,8.29,185.99,66.77,81.90,Alluvial,381,3,1.00,Co0238,498.09,6.96,84.8,12.7,31.1,17.5
8,1433.42,32.41,8.53,179.97,65.92,78.38,Alluvial,378,3,0.92,Co86032,498.00,7.08,90.6,7.4,33.3,17.2
9,1970.00,35.20,9.30,239.26,85.18,112.00,Alluvial,422,3,0.97,Co86032,579.08,6.92,98.0,10.2,44.3,17.1


Since Alluvial soil has the highest frequency with 1,771 records, selecting this group allows focused analysis of rainfall, irrigation, yield, and other agricultural conditions associated with this soil type.

In [8]:
correlation = df[
    ["rainfall_mm", "yield_quintal_per_acre"]
].corr()

correlation

,rainfall_mm,yield_quintal_per_acre
rainfall_mm,1.000000,0.842968
yield_quintal_per_acre,0.842968,1.000000


Correlation analysis was performed to determine the relationship between rainfall and sugarcane yield. The output gives a correlation coefficient of 0.842968, indicating a strong positive relationship between the two variables. This shows that in this dataset, higher rainfall is generally associated with higher sugarcane yield. This relationship is useful for the irrigation advisory use case because rainfall is an important factor when determining irrigation requirements.

In [10]:
normalized_df = df.copy()

cols = [
    "rainfall_mm",
    "avg_temperature_celsius",
    "soil_moisture_percent"
]

normalized_df[cols] = (
    normalized_df[cols] - normalized_df[cols].min()
) / (
    normalized_df[cols].max() - normalized_df[cols].min()
)

normalized_df[cols].head()

,rainfall_mm,avg_temperature_celsius,soil_moisture_percent
0,0.798402,0.817241,0.611801
1,0.415906,0.545977,0.363354
2,0.447632,0.349425,0.577640
3,0.563376,0.703448,0.515528
4,0.269120,0.437931,0.291925


Thus normalization was performed on rainfall, average temperature, and soil moisture to bring the three attributes to a common scale between 0 and 1. The output shows that the original values have been transformed into normalized values . This transformation makes the numerical attributes comparable despite having different units and ranges.

In [11]:
df_left = df[
    ["soil_type", "rainfall_mm", "yield_quintal_per_acre"]
].head(5)

df_right = df[
    ["water_requirement_liters_per_day",
     "irrigation_frequency_per_month"]
].head(5)

joined_df = df_left.join(df_right)

joined_df

,soil_type,rainfall_mm,yield_quintal_per_acre,water_requirement_liters_per_day,irrigation_frequency_per_month
0,Alluvial,1734.13,520.17,15.9,3
1,Alluvial,1286.61,463.89,18.2,3
2,Red Soil,1323.73,343.14,18.5,2
3,Clay Loam,1459.15,508.21,19.6,4
4,Clay Loam,1114.87,433.10,18.6,4


Joining was performed to combine environmental, crop-yield, and irrigation related attributes from two DataFrames using their common row index. The output combines soil type, rainfall, yield, water requirement, and irrigation frequency .

In [12]:
merge_left = df[
    ["soil_type", "rainfall_mm", "yield_quintal_per_acre"]
].head(10).copy()

merge_right = df[
    ["water_requirement_liters_per_day",
     "irrigation_frequency_per_month"]
].head(10).copy()

merge_left["record_id"] = range(1, 11)
merge_right["record_id"] = range(1, 11)

merged_df = pd.merge(
    merge_left,
    merge_right,
    on="record_id"
)

merged_df

,soil_type,rainfall_mm,yield_quintal_per_acre,record_id,water_requirement_liters_per_day,irrigation_frequency_per_month
0,Alluvial,1734.13,520.17,1,15.9,3
1,Alluvial,1286.61,463.89,2,18.2,3
2,Red Soil,1323.73,343.14,3,18.5,2
3,Clay Loam,1459.15,508.21,4,19.6,4
4,Clay Loam,1114.87,433.10,5,18.6,4
5,Alluvial,1418.24,498.09,6,17.5,3
6,Sandy Loam,864.72,283.30,7,19.4,5
7,Sandy Loam,811.86,267.94,8,19.8,5
8,Alluvial,1433.42,498.00,9,17.2,3
9,Alluvial,1970.00,579.08,10,17.1,3


Merging was performed to combine agricultural and irrigation related information using record_id as a common key.

In [13]:
df_part1 = df.iloc[:2500].copy()
df_part2 = df.iloc[2500:].copy()

concatenated_df = pd.concat(
    [df_part1, df_part2],
    axis=0
)

print("Original Dataset Shape:", df.shape)
print("Concatenated Dataset Shape:", concatenated_df.shape)

Original Dataset Shape: (5000, 17)
Concatenated Dataset Shape: (5000, 17)


Concatenation was performed by dividing the dataset into two parts of 2,500 records each and combining them vertically.